keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1458 @@
+downgrade a cluster that has workloads using the pod-level limits functionality, following steps need to be
+performed:
+
+* Translate pod-level resource specifications to container-level specifications.
+
+* Delete existing workloads and recreate them with container-level specifications.
+
+* Drain the nodes one by one to prevent the new pods from being scheduled on them.
+
+* Downgrade the kubelet on the drained nodes to a version that doesn't support pod-level resource specification
+fields.
+
+* Downgrade the control plane components.
+
+* Uncordon the nodes, making it available for scheduling again.
+
+
+### Version Skew Strategy
+
+Pod-level resource specification is an opt-in feature. For this feature to work correctly, it must be enabled
+in all parts of the cluster (scheduler, API server, kubelet). However, if a node is running an older version
+of kubelet that doesn't support this feature, it will ignore the pod-level resouce specifications and continue
+to use the aggregated container-level values (backward compatible). Applications must be prepared for this version
+skew implementation.
+
+For users wanting to use this feature, it is user's responsibility to use the correct version of kube-scheduler,
+kube-apiserver and kubelet otherwise the feature will not work as expected.
+
+## Production Readiness Review Questionnaire
+
+
+### Feature Enablement and Rollback
+
+<!--
+This section must be completed when targeting alpha to a release.
+-->
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+<!--
+Pick one of these and delete the rest.
+
+Documentation is available on [feature gate lifecycle] and expectations, as
+well as the [existing list] of feature gates.
+
+[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
+-->
+
+- [X] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: PodLevelResources
+  - Components depending on the feature gate: kubelet, kube-apiserver, kube-scheduler
+  - Will enabling / disabling the feature require downtime of the control
+    plane? No. `resources` must be set/unset in `spec` at pod-level to enable/disable this feature.
+    The workloads need to restart to disable the feature.
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node? No. Only the workloads need to restart.
+
+###### Does enabling the feature change any default behavior?
+
+No. This feature is guarded by a feature gate, and requires setting Pod level `resource` stanza explicitly.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes. Once the feature gates are disabled, pods created with pod level resource spec need to be deleted and recreated with only container level resources.
```

Comment:
we rarely have a requirement to delete all pods scheduled with this feature enabled.

If we require that - we need to delete those pods first and then disable the feature. It needs to be clearly stated in docs for alpha feature.

Reply:
I don't think we need to mention the recreation. +1 to just disable the feature and delete resources using pod level resource spec.

Reply:
deleting pods automatically is not expected on feature disablement.

It would be OK to say "Pods created with the feature on will retain the resources limits applied by the feature". Generally, we don't expect 
- automatically recreate the pods
- adjust running pods to disable the feature in those running pods.

We would expect that if the pod gets recreated by a deployment or something, that the new pod will not have the feature. But the act of redeploying should be a user decision. The behavior does need to be documented, though.

Some edge cases include - what would be the behavior of kubelet already allocated memory based on pod level resources and when the feature disabled a new Pod was scheduled on that node. Will it result in any issues? Those cases needs to be documented

Reply:
See the reply in another comment: https://github.com/kubernetes/enhancements/pull/4678/files#r1645199565

We can document that there might be a mismatch in resource calculation and actual usage. To fix the issue, users can delete the running pods and create new ones.

That seems sufficient to address most of the issues IMO.

Reply:
+1  to @SergeyKanzhelev here. What I'd like to see is that when the feature gate is disabled, no additional pods may use the feature, but pods that  already exist and are using the feature continue to work as expected.